src/hotspot/share/gc/shenandoah/shenandoahPacer.cpp

```diff
@@ -249,32 +246,25 @@ void ShenandoahPacer::pace_for_alloc(size_t words) {
   JavaThread* current = JavaThread::current();
   if (current->is_attaching_via_jni() ||
       !current->is_active_Java_thread()) {
+    claim_for_alloc<true>(words);
     return;
   }
 
-  double start = os::elapsedTime();
-
-  size_t max_ms = ShenandoahPacingMaxDelay;
-  size_t total_ms = 0;
-
-  while (true) {
+  jlong const max_delay = ShenandoahPacingMaxDelay * NANOSECS_PER_MILLISEC;
+  jlong const start_time = os::elapsed_counter();
+  while (!claimed && (os::elapsed_counter() - start_time) < max_delay) {
     // We could instead assist GC, but this would suffice for now.
-    size_t cur_ms = (max_ms > total_ms) ? (max_ms - total_ms) : 1;
-    wait(cur_ms);
-
-    double end = os::elapsedTime();
-    total_ms = (size_t)((end - start) * 1000);
-
-    if (total_ms > max_ms || Atomic::load(&_budget) >= 0) {
-      // Exiting if either:
-      //  a) Spent local time budget to wait for enough GC progress.
-      //     Breaking out and allocating anyway, which may mean we outpace GC,
-      //     and start Degenerated GC cycle.
-      //  b) The budget had been replenished, which means our claim is satisfied.
-      ShenandoahThreadLocalData::add_paced_time(JavaThread::current(), end - start);
-      break;
-    }
+    wait(1);
+    claimed = claim_for_alloc<false>(words);
+  }
+  if (!claimed) {
+    // Spent local time budget to wait for enough GC progress.
+    // Force allocating anyway, which may mean we outpace GC,
+    // and start Degenerated GC cycle.
+    claimed = claim_for_alloc<true>(words);
+    assert(claimed, "Should always succeed");
   }
+  ShenandoahThreadLocalData::add_paced_time(JavaThread::current(), (double)(os::elapsed_counter() - start_time) / (double) NANOSECS_PER_SEC);
```

Comment:
We already have `current` (`JavaThread::current()`) in scope here, use that :)
I also think a second cast to `(double) NANOSECS_PER_SEC` is redundant.